app/com/arpnetworking/metrics/portal/reports/ReportExecution.java

```diff
@@ -0,0 +1,184 @@
+            final ReportSource source,
+            final Instant scheduled
+    ) {
+        final Map<ReportFormat, RenderedReport> result = Maps.newHashMap(); // TODO(spencerpearson) -- ConcurrentMap?
+        final CompletableFuture<?>[] resultSettingFutures = formats
+                .stream()
+                .map(format -> render(injector, source, format, scheduled)
+                        .thenApply(rendered -> result.put(format, rendered))
+                        .toCompletableFuture())
+                .toArray(CompletableFuture[]::new);
+        return CompletableFuture.allOf(resultSettingFutures).thenApply(nothing -> ImmutableMap.copyOf(result));
+    }
+
+    /* package private */ static <S extends ReportSource, F extends ReportFormat> CompletionStage<RenderedReport> render(
+            final Injector injector,
+            final S source,
+            final F format,
+            final Instant scheduled
+    ) {
+        return getRenderer(injector, source, format).render(source, format, scheduled);
+    }
+
+    /* package private */ static CompletionStage<Void> sendAll(
+            final Injector injector,
+            final ImmutableMultimap<Recipient, ReportFormat> recipientToFormats,
+            final ImmutableMap<ReportFormat, RenderedReport> formatToRendered
+    ) {
+        final CompletableFuture<?>[] futures = recipientToFormats
+                .asMap()
+                .entrySet()
+                .stream()
+                .map(entry -> getSender(injector, entry.getKey())
+                        .send(entry.getKey(), mask(formatToRendered, entry.getValue()))
+                        .toCompletableFuture())
+                .toArray(CompletableFuture[]::new);
+        return CompletableFuture.allOf(futures);
+    }
+
+    private static <S extends ReportSource, F extends ReportFormat> Renderer<S, F> getRenderer(
+            final Injector injector,
+            final S source,
+            final F format
+    ) {
+        final String keyName = getRendererKeyName(source, format);
+        @SuppressWarnings("unchecked")
+        final Renderer<S, F> renderer = injector.getInstance(Key.get(Renderer.class, Names.named(keyName)));
+        if (renderer == null) {
+            throw new IllegalArgumentException("no Renderer exists for key name '" + keyName + "'");
+        }
+        return renderer;
+    }
+
+    private static Sender getSender(final Injector injector, final Recipient recipient) {
+        final String keyName = getSenderKeyName(recipient);
+        final Sender sender = injector.getInstance(Key.get(Sender.class, Names.named(keyName)));
+        if (sender == null) {
+            throw new IllegalArgumentException("no Sender exists for key name '" + keyName + "'");
+        }
+        return sender;
+    }
+
+    /* package private */ static String getSenderKeyName(final Recipient recipient) {
+        return CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.LOWER_UNDERSCORE, recipient.getType().name());
```

Comment:
I would prefer that we just had a member or method on `RecipientType` that returned or computed the key for a given type.